fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryIndexPlan.java

```diff
@@ -84,37 +100,103 @@ public class RecordQueryIndexPlan implements RecordQueryPlanWithNoChildren, Reco
+                                 @Nonnull final RecordQueryPlannerConfiguration.IndexPrefetchUse useIndexPrefetch,
                                  final boolean reverse,
                                  final boolean strictlySorted,
                                  @Nonnull final Optional<? extends ScanWithFetchMatchCandidate> matchCandidateOptional,
                                  @Nonnull final Type resultType) {
         this.indexName = indexName;
+        this.commonPrimaryKey = commonPrimaryKey;
         this.scanParameters = scanParameters;
+        this.useIndexPrefetch = useIndexPrefetch;
         this.reverse = reverse;
         this.strictlySorted = strictlySorted;
         this.matchCandidateOptional = matchCandidateOptional;
         this.resultType = resultType;
+        if (useIndexPrefetch != RecordQueryPlannerConfiguration.IndexPrefetchUse.NONE) {
+            if (commonPrimaryKey == null) {
+                error("Index Prefetch cannot be used without a primary key. Falling back to regular scan.");
+                this.useIndexPrefetch = RecordQueryPlannerConfiguration.IndexPrefetchUse.NONE;
+            }
+            if (scanParameters.getScanType() != IndexScanType.BY_VALUE) {
+                error("Index Prefetch can only be used with VALUE index scan. Falling back to regular scan.");
+                this.useIndexPrefetch = RecordQueryPlannerConfiguration.IndexPrefetchUse.NONE;
+            }
+            if (!FDBDatabaseFactory.instance().getDatabase().getAPIVersion().isAtLeast(APIVersion.API_VERSION_7_1)) {
+                error("Index Prefetch can only be used with API_VERSION of at least 7.1. Falling back to regular scan.");
+                this.useIndexPrefetch = RecordQueryPlannerConfiguration.IndexPrefetchUse.NONE;
+            }
+        }
+    }
+
+    @Nonnull
+    @Override
+    public <M extends Message> RecordCursor<QueryResult> executePlan(@Nonnull final FDBRecordStoreBase<M> store, @Nonnull final EvaluationContext context,
+                                                                     @Nullable final byte[] continuation, @Nonnull final ExecuteProperties executeProperties) {
+        switch (useIndexPrefetch) {
+            case NONE:
+                // Use the default implementation for index scan
+                return RecordQueryPlanWithIndex.super.executePlan(store, context, continuation, executeProperties);
+            case USE_INDEX_PREFETCH:
+                // Use index prefetch without fallback
+                return executeUsingIndexPrefetch(store, context, continuation, executeProperties);
+            case USE_INDEX_PREFETCH_WITH_FALLBACK:
+                // Use Index prefetch and fall back to regular index scan.
+                // Using the fallback mechanism here separates the execution part from the planning part
+                // (No need to plan again) and from failures at other parts of the execution.
```

Comment:
This comment is misleading. Please ignore. The continuations are assumed to be compatible between execution modes.